### PR TITLE
Enable retrieving both transactions and blocks in json format

### PR DESCRIPTION
**Pull request by TheQuantumPhysicist**

### CODE REVIEW

**Review by iljakuklic — COMMENTED** [review]
Just gave it a quick first pass.

Besides the minor comments below, one more thought regarding the serialization of binary data. Binary data should really be encoded using `encode_bytes` rather than `encode_str`. However, I suspect serde-json encodes bytes as an json array of numbers. That could be addressed by creating a wrapper serializer that forwards `encode_bytes` to `encode_str` by converting it into hex and passes through everything else into the inner serializer. One shortcoming is that convenience methods like `serde_json::to_string` will no longer be available since they hard-code the serializer from the crate. Thoughts?

File: common/src/chain/transaction/signature/inputsig/mod.rs [hunk 1]
```diff
@@ -29,3 +29,9 @@ pub enum InputWitness {
     #[codec(index = 1)]
     Standard(StandardInputSignature),
 }
+
+impl serde::Serialize for InputWitness {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        serializer.serialize_str(&self.hex_encode())
```
[the remark this answers is not in the document]
Comment (TheQuantumPhysicist) [reply common/src/chain/transaction/signature/inputsig/mod.rs:35]: Yep. It's to make that distinction.

Done that serialization thing.

File: crypto/Cargo.toml [hunk 2]
```diff
@@ -28,6 +28,8 @@ parity-scale-codec.workspace = true
 rand.workspace = true
 ripemd.workspace = true
 schnorrkel.workspace = true
+serde = { workspace = true, features = ["derive"] }
+serde_json.workspace = true
```
[the remark this answers is not in the document]
Comment (TheQuantumPhysicist) [reply crypto/Cargo.toml:32]: It was being used until I realized the array of numbers problem.

File: serialization/src/json_encoded.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    pub fn take(self) -> T {
+        self.0
+    }
+}
+
+impl<'de, T: serde::Deserialize<'de>> JsonEncoded<T> {
+    // We cannot use FromStr because of the lifetime limitation
+    pub fn from_string(s: &'de str) -> Result<Self, serde_json::Error> {
+        serde_json::from_str(s)
+    }
```
[the remark this answers is not in the document]
Comment (TheQuantumPhysicist) [reply serialization/src/json_encoded.rs:36]: Not sure what to call it... from_str is reserved.

File: serialization/src/json_encoded.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        Self(value)
+    }
+
+    pub fn take(self) -> T {
+        self.0
+    }
+}
+
+impl<'de, T: serde::Deserialize<'de>> JsonEncoded<T> {
+    // We cannot use FromStr because of the lifetime limitation
```
[the remark this answers is not in the document]
Comment (TheQuantumPhysicist) [reply serialization/src/json_encoded.rs:33]: Nice idea. Done.

File: serialization/src/json_encoded.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (iljakuklic) [reply serialization/src/json_encoded.rs:36]: > Not sure what to call it... from_str is reserved.

Ok, that's annoying. Maybe something like `new_from_str` or `create_from_str` but that breaks the usual naming conventions. I don't have a good suggestion I'm afraid.

[the remark this answers is not in the document]
Comment (TheQuantumPhysicist) [reply serialization/src/json_encoded.rs:36]: from_string_slice...

**Review by iljakuklic — COMMENTED** [review]
Still somewhat puzzling over the `from_string` name. Good otherwise.

File: serialization/src/json_encoded.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (iljakuklic) [reply serialization/src/json_encoded.rs:36]: > from_string_slice...

Yea that's acceptable

**Review by iljakuklic — APPROVED** [review]

**Review by muursh — APPROVED** [review]

**Pull request merged**